oeps/oep-0019-bp-django-caches.rst

```diff
@@ -0,0 +1,228 @@
+~~~~~~~~~~~~~~~~~~~~~
+
+Either use a cache decorator that generates the cache keys, or use a trusted
+library for generating cache keys that will be Memcached safe and will avoid
+key conflict in the global space in Memcached.
+
+See `Implementations`_ for more details.
+
+Implementations
+~~~~~~~~~~~~~~~
+
+This section will be updated as a shared library is selected and evolved.
+
+**In edx-platform:**
+
+* There is no utility for handling the `Common Caching Defect and Fix`_.
+
+* There is a RequestCache, but no Tiered-Cache.
+
+* There are no utilities for forcing cache misses.
+
+* There are various cache decorators. These could be assessed when considering
+  creating a shared library of caching utilities.
+
+* There are common utilities for generating cache keys. These should be be
+  assessed when considering creating a shared library of caching utilities.
+
+**In ecommerce:**
+
+* The `RequestCache and TieredCache`_ both handle the `Common Caching Defect and
+  Fix`_. Although it is simple to refactor standard Django caching code to use
+  these classes, the interface to these classes is non-standard and they
+  the simple refactor doesn't move toward using cache decorators.
+
+* There is both a `RequestCache and TieredCache`_.
+
+* The TieredCache provides functionality for forcing cache misses.
+
+* There is no cache decorator utility.
+
+* There are common utilities for generating cache keys. These should be be
+  assessed when considering creating a shared library of caching utilities.
+
+**Quickcache library**
+
+Note: The `quickcache library`_ has not been fully assessed. It is provided as a
+good potential candidate for a shared library.
+
+* The `Common Caching Defect and Fix`_ is handled by using a decorator. Any calls
+  to the Django cache outside of the decorator would be susceptible to the bug.
+
+* Provides a TieredCache, but no RequestCache. The library would need to be
+  enhanced to include the RequestCache, or to extend to use one.
+
+* There is functionality for skipping the cache. However, this functionality still
+  needs to be assessed.
+
+* The library is based on a slick @quickcache decorator. The decorator is
+  opinionated about its cache keys, tied to the code it is wrapping. This needs
+  to be assessed to see if it is a good fit.
+
+* The cache key generation is built into the decorator. Again, this needs to be
+  assessed.
+
+**Next steps:**
+
+* Choose the best solution(s) to the above problems and create a shared library
+  (or libraries) that enable these best practices.
+
+* Use linting utilities to enforce the best practices.
+
+.. RequestCache and TieredCache: https://github.com/edx/ecommerce/blob/master/ecommerce/cache_utils/README.rst
+.. quickcache library: https://github.com/dimagi/quickcache
+
+Consequences
+------------
+
+As long as we keep all options open for developers, these best practices and
+supporting libraries should only help developers right cleaner and less buggy
```

Comment:
s/right/write/

Reply:
The brain works (or malfunctions) in mysterious ways.